File: chirag/frontend/src/lib/schemas/voice.ts
import { z } from "zod";

import { ChatStepSchema } from "@/lib/schemas/chat";

export const VoiceTimingsSchema = z.object({
  stt_ms: z.number().int().nonnegative().default(0),
  translate_ms: z.number().int().nonnegative().default(0),
  answer_ms: z.number().int().nonnegative().default(0),
  humanize_ms: z.number().int().nonnegative().default(0),
  backtranslate_ms: z.number().int().nonnegative().default(0),
  total_ms: z.number().int().nonnegative().default(0),
});
export type VoiceTimings = z.infer<typeof VoiceTimingsSchema>;

export const VoiceRespondResponseSchema = z.object({
  transcript: z.string(),
  detected_language: z.string().default(""),
  detected_language_name: z.string().default(""),
  language_probability: z.number().min(0).max(1).default(0),
  english_transcript: z.string().default(""),
  answer_raw: z.string(),
  answer_spoken_en: z.string().default(""),
  answer_spoken: z.string(),
  answer_language: z.string().default("en"),
  steps: z.array(ChatStepSchema).default([]),
  timings: VoiceTimingsSchema.default({
    stt_ms: 0,
    translate_ms: 0,
    answer_ms: 0,
    humanize_ms: 0,
    backtranslate_ms: 0,
    total_ms: 0,
  }),
  llm_model: z.string(),
  voice_id: z.string(),
  schema_version: z.string(),
});
export type VoiceRespondResponse = z.infer<typeof VoiceRespondResponseSchema>;

export const VoiceConfigSchema = z.object({
  ready: z.boolean(),
  voice_id: z.string(),
  schema_version: z.string(),
});
export type VoiceConfig = z.infer<typeof VoiceConfigSchema>;

File: chirag/src/agnes/models/voice.py
"""Pydantic contracts for the /api/voice/* endpoints."""

from __future__ import annotations

from pydantic import BaseModel, Field

from agnes.models.chat import ChatStep

VOICE_SCHEMA_VERSION = "voice-1.1.0"


class VoiceTimings(BaseModel):
    """Per-stage latency breakdown (milliseconds)."""

    stt_ms: int = 0
    translate_ms: int = 0
    answer_ms: int = 0
    humanize_ms: int = 0
    backtranslate_ms: int = 0
    total_ms: int = 0


class VoiceRespondResponse(BaseModel):
    """Answer the frontend gets *before* it asks for TTS audio."""

    transcript: str = Field(
        ...,
        description="User speech in the original spoken language (as recognised by Scribe v1).",
    )
    detected_language: str = Field(
        default="",
        description="ISO 639-1 code detected by Scribe (e.g. 'en', 'fr', 'de'). Empty if unknown.",
    )
    detected_language_name: str = Field(
        default="",
        description="Human-readable name of the detected language (e.g. 'English').",
    )
    language_probability: float = Field(
        default=0.0,
        description="Scribe confidence (0..1) for the detected language.",
    )
    english_transcript: str = Field(
        default="",
        description=(
            "English paraphrase produced by the translator agent. Equals "
            "`transcript` when the user already spoke English."
        ),
    )
    answer_raw: str = Field(
        ...,
        description=(
            "Raw reply from the answer agent (English, may contain markdown / tables)."
        ),
    )
    answer_spoken_en: str = Field(
        default="",
        description="Humanised English spoken reply (always produced).",
    )
    answer_spoken: str = Field(
        ...,
        description=(
            "Final speech-ready reply in the user's spoken language. Equal to "
            "`answer_spoken_en` for English users; otherwise back-translated."
        ),
    )
    answer_language: str = Field(
        default="en",
        description="ISO 639-1 code of `answer_spoken` (matches `detected_language` for non-English users).",
    )
    steps: list[ChatStep] = Field(default_factory=list)
    timings: VoiceTimings = Field(default_factory=VoiceTimings)
    llm_model: str
    voice_id: str
    schema_version: str = VOICE_SCHEMA_VERSION


class VoiceTtsRequest(BaseModel):
    text: str = Field(..., min_length=1, max_length=4000)
    voice_id: str | None = Field(
        default=None,
        description="Optional ElevenLabs voice id override (default = Charlotte).",
    )


__all__ = [
    "VOICE_SCHEMA_VERSION",
    "VoiceRespondResponse",
    "VoiceTimings",
    "VoiceTtsRequest",
]
